pkg/api/bzzfile.go

```diff
@@ -0,0 +1,245 @@
+	}
+}
+
+// bzzFileDownloadHandler downloads the file given the entry's reference.
+func (s *server) bzzFileDownloadHandler(w http.ResponseWriter, r *http.Request) {
+	addr := mux.Vars(r)["addr"]
+	address, err := swarm.ParseHexAddress(addr)
+	if err != nil {
+		s.Logger.Debugf("bzz-file: parse file address %s: %v", addr, err)
+		s.Logger.Error("bzz-file: parse file address")
+		jsonhttp.BadRequest(w, "invalid file address")
+		return
+	}
+
+	// read entry.
+	j := joiner.NewSimpleJoiner(s.Storer)
+	buf := bytes.NewBuffer(nil)
+	_, err = file.JoinReadAll(j, address, buf)
+	if err != nil {
+		s.Logger.Debugf("bzz-file: read entry %s: %v", addr, err)
+		s.Logger.Error("bzz-file: read entry")
+		jsonhttp.InternalServerError(w, "error reading entry")
+		return
+	}
+	e := &entry.Entry{}
+	err = e.UnmarshalBinary(buf.Bytes())
+	if err != nil {
+		s.Logger.Debugf("bzz-file: unmarshall entry %s: %v", addr, err)
+		s.Logger.Error("bzz-file: unmarshall entry")
+		jsonhttp.InternalServerError(w, "error unmarshalling entry")
+		return
+	}
+
+	// If none match header is set always send the reply as not modified
+	// TODO: when SOC comes, we need to revisit this concept
+	noneMatchEtag := r.Header.Get("If-None-Match")
+	if noneMatchEtag != "" {
+		if e.Reference().Equal(address) {
+			w.WriteHeader(http.StatusNotModified)
+			return
+		}
+	}
+
+	// Read metadata.
+	buf = bytes.NewBuffer(nil)
+	_, err = file.JoinReadAll(j, e.Metadata(), buf)
+	if err != nil {
+		s.Logger.Debugf("bzz-file: read metadata %s: %v", addr, err)
+		s.Logger.Error("bzz-file: read netadata")
+		jsonhttp.InternalServerError(w, "error reading metadata")
+		return
+	}
+	metaData := &entry.Metadata{}
+	err = json.Unmarshal(buf.Bytes(), metaData)
+	if err != nil {
+		s.Logger.Debugf("bzz-file: unmarshall metadata %s: %v", addr, err)
+		s.Logger.Error("bzz-file: unmarshall metadata")
+		jsonhttp.InternalServerError(w, "error unmarshalling metadata")
+		return
+	}
+
+	// send the file data back in the response
+	outBuffer := bytes.NewBuffer(nil)
+	_, err = file.JoinReadAll(j, e.Reference(), outBuffer)
+	if err != nil {
+		s.Logger.Debugf("bzz-file: data read %s: %v", addr, err)
+		s.Logger.Error("bzz-file: data read")
+		jsonhttp.InternalServerError(w, "error reading data")
+		return
+	}
+	w.Header().Set("ETag", fmt.Sprintf("%q", e.Reference()))
+	w.Header().Set("Content-Disposition", fmt.Sprintf("inline; filename=\"%s\"", metaData.Filename))
+	w.Header().Set("Content-Type", metaData.MimeType)
+	_, _ = io.Copy(w, outBuffer)
+}
+
+// storeMeta is used to store metadata information as a whole.
+func (s *server) storeMeta(ctx context.Context, dataBytes []byte) (swarm.Address, error) {
+	dataBuf := bytes.NewBuffer(dataBytes)
+	dataReader := io.LimitReader(dataBuf, int64(len(dataBytes)))
+	dataReadCloser := ioutil.NopCloser(dataReader)
+	o, err := s.splitUpload(ctx, dataReadCloser, int64(len(dataBytes)))
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	bytesResp := o.(bytesPostResponse)
```

Comment:
This is very unfortunate. But this is a consequence of a badly written splitUpload. We can address this later in the PR when we address technical dept created in the pr that added splitUpload.